### PR TITLE
[Parallel] SQuAD multi-GPU script

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
@szhengac 

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by szhengac** [comment]
I would recommend using horovod for finetuning task at this time, as the
pre-built newest mxnet does not work with horovod at the moment. The user
has to use older version of mxnet or compile it from source.

**Comment by leezu** [comment]
Nightly builds are fixed and work with horovod since 2 weeks. See https://github.com/apache/incubator-mxnet/issues/16193

**Comment by eric-haibin-lin** [comment]
@szha @ptrendx mxnet pypi limit was increased. Will mxnet 1.6 stable release include the symbols needed by horovod?

**Comment by szha** [comment]
@eric-haibin-lin yes

File: scripts/bert/finetune_squad.py [hunk 1]
```diff
@@ -376,62 +404,81 @@ def set_new_lr(step_num, batch_id):
...
                 if accumulate:
-                    ls = ls / accumulate
-            ls.backward()
+                    loss = loss / accumulate
+                if args.dtype == 'float16':
+                    with amp.scale_loss(loss, trainer) as l:
+                        mx.autograd.backward(l)
+                        norm_clip = 1.0 * size * trainer._amp_loss_scaler.loss_scale
+                else:
+                    mx.autograd.backward(l)
```
Comment (xinyu-intel) [inline scripts/bert/finetune_squad.py]: ```mx.autograd.backward(loss)```

Comment (eric-haibin-lin) [reply scripts/bert/finetune_squad.py]: thanks for pointing it out!

File: scripts/bert/finetune_squad.py [hunk 2]
```diff
@@ -229,10 +251,10 @@
 batch_size = args.batch_size
 test_batch_size = args.test_batch_size
 lr = args.lr
-ctx = mx.cpu() if args.gpu is None else mx.gpu(args.gpu)
+ctx = mx.gpu(rank) if args.gpu else mx.cpu()
```
Comment (xinyu-intel) [inline scripts/bert/finetune_squad.py]: suggest using `hvd.local_rank()` to support multi gpu instance.

**Comment by mli** [comment]
Job PR-1079/16 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1079/16/index.html

**Comment by mli** [comment]
Job PR-1079/17 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1079/17/index.html

File: scripts/bert/finetune_squad.py [hunk 3]
```diff
@@ -215,6 +225,26 @@
...
+    rank = 0
+    size = 1
+    local_rank = 0
+
+if args.dtype == 'float16':
+    # patch AMP due to issue: https://github.com/apache/incubator-mxnet/issues/17409
+    from mxnet.contrib import amp
+    ops = ['_contrib_interleaved_matmul_encdec_qk', '_contrib_interleaved_matmul_encdec_valatt',
+           '_contrib_interleaved_matmul_selfatt_qk', '_contrib_interleaved_matmul_selfatt_valatt']
+    amp.lists.symbol.WIDEST_TYPE_CASTS.extend(ops)
```
Comment (szhengac) [inline scripts/bert/finetune_squad.py]: What is this?

Comment (eric-haibin-lin) [reply scripts/bert/finetune_squad.py]: A bug in AMP. But the fix was merged to mxnet nightly, so I'll just remove these temporary fixes.

**Comment by mli** [comment]
Job PR-1079/18 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1079/18/index.html

File: scripts/bert/finetune_squad.py (hunk 3 above)
Comment (leezu) [reply scripts/bert/finetune_squad.py]: Backported to 1.6 release?

**Comment by mli** [comment]
Job PR-1079/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1079/19/index.html

File: scripts/bert/finetune_squad.py [hunk 4]
```diff
@@ -335,28 +360,34 @@ def train():
     log.info('The number of examples after preprocessing:{}'.format(
         len(train_data_transform)))
 
+    sampler = nlp.data.SplitSampler(len(train_data_transform), num_parts=size,
+                                    part_index=rank, even_size=True)
```
Comment (szhengac) [inline scripts/bert/finetune_squad.py]: `even_size=True`? It is possible that the samples cannot be evenly split.

File: scripts/bert/finetune_squad.py [hunk 5]
```diff
@@ -376,62 +407,82 @@ def set_new_lr(step_num, batch_id):
...
     log_num = 0
-    for epoch_id in range(epochs):
-        step_loss = 0.0
-        tic = time.time()
-        for batch_id, data in enumerate(train_dataloader):
+    batch_id = 0
+    step_loss = 0.0
+    tic = time.time()
+    step_num = 0
+    while step_num < num_train_steps:
```
Comment (szhengac) [inline scripts/bert/finetune_squad.py]: Not number of epochs?

Comment (eric-haibin-lin) [reply scripts/bert/finetune_squad.py]: num_train_steps is calculated based on num_epochs

File: scripts/bert/finetune_squad.py (hunk 4 above)
Comment (eric-haibin-lin) [reply scripts/bert/finetune_squad.py]:
```
    even_size: bool, default False
      If the number of samples is not even across all partitions, sample a few extra samples
      for the ones with fewer samples.
```

File: scripts/bert/finetune_squad.py (hunk 3 above)
Comment (eric-haibin-lin) [reply scripts/bert/finetune_squad.py]: yes

File: scripts/bert/finetune_squad.py (hunk 4 above)
Comment (szhengac) [reply scripts/bert/finetune_squad.py]: Then, it is possible that some workers do not have gradient to contribute at the end of each epoch.

File: scripts/bert/finetune_squad.py (hunk 5 above)
Comment (szhengac) [reply scripts/bert/finetune_squad.py]: You training for `num_train_steps * len(train_dataloader) = num_epoch * num_steps_per_epoch * len(train_dataloader)` iterations?

File: scripts/bert/finetune_squad.py (hunk 4 above)
Comment (eric-haibin-lin) [reply scripts/bert/finetune_squad.py]: It means that the sampler will duplicate with extra samples if the size is not even

Comment (szhengac) [reply scripts/bert/finetune_squad.py]: ok, i just saw the `extras` in the `__iter__`.

**Comment by mli** [comment]
Job PR-1079/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1079/20/index.html

**Comment by mli** [comment]
Job PR-1079/21 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1079/21/index.html

File: scripts/bert/finetune_squad.py [hunk 6]
```diff
@@ -391,71 +418,89 @@ def set_new_lr(step_num, batch_id):
...
+                        norm_clip = 1.0 * size * trainer._amp_loss_scaler.loss_scale
+                else:
+                    mx.autograd.backward(loss)
+                    norm_clip = 1.0 * size
+
             # update
             if not accumulate or (batch_id + 1) % accumulate == 0:
                 trainer.allreduce_grads()
-                nlp.utils.clip_grad_global_norm(params, 1)
+                nlp.utils.clip_grad_global_norm(params, norm_clip)
```
Comment (szhengac) [inline scripts/bert/finetune_squad.py]: if it was clipped, we need to set `optimizer.rescale_grad` to 1.
